sway/desktop/output.c

```diff
@@ -224,11 +224,146 @@ static void render_view(struct sway_view *view, struct sway_output *output) {
+		wlr_backend_get_renderer(output->wlr_output->backend);
+	struct wlr_box box;
+	float color[4];
+	size_t num_tabs = 1;
+	size_t tab_width = 1;
+	
+	if (con->parent->layout == L_TABBED) {
+		num_tabs = con->parent->children->length;
+		tab_width = (con->width / num_tabs);
+	}
+
+	// Single pixel bar above title
+	memcpy(&color, colors->border, sizeof(float) * 4);
+	color[3] *= con->alpha;
+	box.x = con->x + depth * tab_width;
+	box.y = con->y;
+	box.width = con->width / num_tabs;
+	box.height = 1;
+	scale_box(&box, output->wlr_output->scale);
+	wlr_render_rect(renderer, &box, color,
+			output->wlr_output->transform_matrix);
+
+	// Single pixel bar below title
+	box.x = con->x + depth * tab_width;
+	box.y = con->sway_view->y - 1;
+	box.width = con->width / num_tabs;
+	box.height = 1;
+	scale_box(&box, output->wlr_output->scale);
+	wlr_render_rect(renderer, &box, color,
+			output->wlr_output->transform_matrix);
+
+	// Title background
+	memcpy(&color, colors->background, sizeof(float) * 4);
+	color[3] *= con->alpha;
+	box.x = con->x + depth * tab_width;
+	box.y = con->y + 1;
+	box.width = con->width / num_tabs + 2;
+	box.height = con->sway_view->y - con->y - 2;
+	scale_box(&box, output->wlr_output->scale);
+	wlr_render_rect(renderer, &box, color,
+			output->wlr_output->transform_matrix);
+
+	// Title text
+	if (title_texture) {
+		wlr_renderer_scissor(renderer, &box);
+		wlr_render_texture(renderer, title_texture,
+				output->wlr_output->transform_matrix, box.x + 2, box.y, 1);
+		wlr_renderer_scissor(renderer, NULL);
+	}
+}
+
+/**
+ * Render decorations for the top border for tabbed view with "border normal".
+ */
+static void render_container_top_tabbed_border_normal(struct sway_output *output,
+		struct sway_container *con, struct wlr_texture **title_texture,
+		int focus_number, bool active) {
+
+	int index_tabs = con->parent->children->length - 1;
+	int index_left_of_focus = focus_number - 1;
+	struct border_colors *unfocused = select_focus_color("unfocused");
+	struct border_colors *focus_color;
+
+	if (active) {
+		focus_color = select_focus_color("focused");
+	} else {
+		focus_color = select_focus_color("inactive");
+	}
+
+	if (index_left_of_focus >= 0) {
+		for (int i = 0; i <= index_left_of_focus; ++i) {
+			render_container_top_border_normal(output, con, unfocused, title_texture[i], i);
+		}
+	}
+	render_container_top_border_normal(output, con, focus_color, title_texture[focus_number], focus_number);
+	if (index_tabs > index_left_of_focus) {
+		for (int j = focus_number + 1; j <= index_tabs; ++j) {
+			render_container_top_border_normal(output, con, unfocused, title_texture[j], j);
+		}
+	}
+}
+
+/**
+ * Render decorations for a horizontal or vertical view with "border normal".
  */
 static void render_container_simple_border_normal(struct sway_output *output,
 		struct sway_container *con, struct border_colors *colors,
-		struct wlr_texture *title_texture) {
+		struct wlr_texture *title_texture, bool top) {
```

Comment:
To be honest I'm not a fan of piggybacking off this function. "simple" refers to a linear layout, being L_HORIZ or L_VERT.

It's a tradeoff between code duplication vs the complexity of having separate execution paths through the function. I prefer the code duplication (separate function) option.